libraries/libfc/include/fc/container/ordered_diff.hpp

```diff
@@ -24,87 +34,138 @@ template <typename T, typename SizeType = size_t, template<typename Y, typename.
+               result.move_indexes.emplace_back(orig_sorted[orig_index].second, targ_sorted[cur_index].second);
             }
-         } else if (s < source.size()) {
-            // remove extra in source
-            assert(s <= std::numeric_limits<SizeType>::max());
-            result.remove_indexes.push_back(s);
-            ++s;
-         } else if (t < target.size()) {
-            // insert extra in target
-            assert(t <= std::numeric_limits<SizeType>::max());
-            result.insert_indexes.emplace_back(t, target[t]);
-            ++t;
+            ++orig_index;
+            ++cur_index;
          }
       }
 
+      std::ranges::sort(result.remove_indexes, [](auto&& lhs, auto&& rhs) { return lhs > rhs; }); // descending
+      std::ranges::sort(result.insert_indexes, [](auto&& lhs, auto&& rhs) { return lhs.first < rhs.first; });
+      std::ranges::sort(result.move_indexes,   [](auto&& lhs, auto&& rhs) { return lhs.first < rhs.first; });
+
+      // update move indexes
+      if (!result.move_indexes.empty()) {
+         // remove decreases `from` index
+         for (const auto& i : result.remove_indexes) {
+            auto m = std::lower_bound(result.move_indexes.begin(), result.move_indexes.end(), i, [&](const auto& p, const auto& v) { return p.first < v; });
+            for (; m != result.move_indexes.end(); ++m) {
+               --m->first;
+            }
+         }
+         // insert increases `from` index
+         for (const auto& i : result.insert_indexes) {
+            auto m = std::lower_bound(result.move_indexes.begin(), result.move_indexes.end(), i.first, [](const auto& p, const auto& v) { return p.first < v; });
+            for (; m != result.move_indexes.end(); ++m) {
+               ++m->first;
+            }
+         }
+         // remove any moves that are not needed, from == to
+         result.move_indexes.erase(std::remove_if(result.move_indexes.begin(), result.move_indexes.end(),
+                                                  [](const auto& p) { return p.first == p.second; }),
+                                   result.move_indexes.end());
+      }
+
       return result;
    }
 
-   /// @param diff the diff_result created from diff(source, target), apply_diff(std::move(source), diff_result) => target
-   /// @param container the source of diff(source, target) to modify using the diff_result to produce original target
-   /// @return the modified container now equal to original target
+   /// @param diff the diff_result created from diff(orig, target), apply_diff(std::move(orig), diff_result) => target
+   /// @param container the orig of diff(orig, target) to modify using the diff_result to produce the target
+   /// @return the modified container now equal to target
    template <typename X>
    requires std::same_as<std::decay_t<X>, diff_result>
    static Container<T> apply_diff(Container<T>&& container, X&& diff) {
```

Comment:
Should `container` be sorted first?

It is also better to be named `original` or `orig`.